File: letsbuild/srcs/Interpreter.cpp
#include "./Interpreter.hpp"

//CONSTRUCTORS
Interpreter::Interpreter()
: _parser(NULL) {}

Interpreter::Interpreter(Parser &parser)
: _parser(&parser) {}

Interpreter::Interpreter(Parser const &src)
{
	*this = src;
}

Interpreter::~Interpreter() {}


//OVERLOADS
Interpreter	&Interpreter::operator=(Interpreter const &rhs)
{
	this->_parser = &rhs.getParser();
	return (*this);
};


//GETTERS
Parser	&Interpreter::getParser()	const
{
	return (*_parser);
}

//SETTERS
void	Interpreter::setParser(Parser &parser)
{
	_parser = &parser;
}


//METHODS
void	Interpreter::calcul(Node *node)
{
	if (node->left)
		calcul(node->left);
	if (node->right)
		calcul(node->right);
	
	if (node->data.getType() == PLUS)	
		node->data.setValue((node->left->data.getValue() + node->right->data.getValue()));
    else if (node->data.getType() == MINUS)	
		node->data.setValue((node->left->data.getValue() - node->right->data.getValue()));
    else if (node->data.getType() == MUL)	
		node->data.setValue((node->left->data.getValue() * node->right->data.getValue()));
    else if (node->data.getType() == DIV)	
		node->data.setValue((node->left->data.getValue() / node->right->data.getValue()));
}

Number	Interpreter::interpret()
{
	Node	*ast = _parser->parse();
	if (_parser->getCurrent_token().getType() != EOL)
		_parser->error();
	this->calcul(ast);
	Number	result = ast->data.getValue();
	cout << *ast << endl;
	clear_btree(ast);
	return (result);
}

File: python_computorv1/srcs/Interpreter.py
from srcs.utils import *
from srcs.BTree import *

class Interpreter(object):
    def __init__(self, parser):
        self.parser = parser
        self.tree = None
        self.steps = ""
    
    def calcul(self, root):
        if root.left is not None:
            self.calcul(root.left)
        if root.right is not None:
            self.calcul(root.right)
        


        if root.data.type == PLUS:
            root.data.value = root.left.data.value + root.right.data.value
        elif root.data.type == MINUS:
            root.data.value = root.left.data.value - root.right.data.value
        elif root.data.type == MUL:
            root.data.value = root.left.data.value * root.right.data.value
        elif root.data.type == DIV:
            root.data.value = root.left.data.value / root.right.data.value
        elif root.data.type == POWER:
            root.data.value = root.left.data.value ** root.right.data.value
        elif root.data.type == EQUAL:
            self.steps += f"-> {root.left.data.value} - ({root.right.data.value}) = 0\n"
            self.steps += f"-> {root.left.data.value - root.right.data.value} = 0\n"
            root.data.value = root.left.data.value - root.right.data.value
        else:
            return
        if root.data.type != EQUAL:
            self.steps += f"-> {str_expr(self.tree)}\n"
        

    def interpret(self):
        self.tree = self.parser.parse()
        self.steps += f"-> {str_expr(self.tree)}\n"
        self.calcul(self.tree)
        return self.tree
